core/utils/STree.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Optional
+
+
+@dataclass
+class Node:
+    s_link: Optional[Node] = None
+    children: dict[str, Node] = field(default_factory=dict)
+    start: int = -1
+    depth: int = -1
+    parent: Optional[Node] = None
+
+    def __str__(self) -> str:
+        return f"children={list(self.children.keys())}"
+
+
+def create_node(s: str, u: Node, depth: int) -> Node:
```

Comment:
This function does more than creating a node. It looks mostly like it is breaking an edge with a new node. I would change the name.

You can simplify it a lot if you use the constructor in `Node`. Something like this:

```python
def create_node(s: str, u: Node, depth: int) -> Node:
     v = Node(u.start, u.depth, u.parent) ## assuming you fix the constructor
     v.children[s[u.start + u.depth]] = u
     u.parent = v
     parent.children[s[u.start + parent.depth]] = v
     return v
 ```

I don't understand this bit: `s[u.start + parent.depth]`. If `start` and `depth` are indices into the underlying string, then adding them makes no sense. You cannot meaningfully add indices, only indices and offsets. If `depth` is an offset or length of a kind, then it doesn't help me either, since the length of the parent's edge shouldn't matter for the out-edge of `u`. It might do the right thing, but I have no idea what it is doing.

If I, in a top-down search, get to a leaf node, and I have some substring `p[k:]` to search for, how do I match against the final edge if all I know is where the suffix `i` starts and how long it is? I don't know which suffix of suffix `i` I am looking at when I consider the leaf-edge.

Reply:
To handle this case, I'm using python's `startswith(prefix)` to match when I'm at the leaf.

Reply:
And I agree with the simplification. It makes sense to reduce the code and just use the constructor(fixed that as well).